### PR TITLE
[Role Datums] Dynamic Mode

### PR DESCRIPTION
**Pull request by DeityLink**
It's not finished but I'm making a PR now so fellow datum smashers can start getting acquainted with the tool they'll be using to spawn our beloved antags into the round.

Only the roundstart portion is coded right now, I'll do latejoin and midround rulesets tomorrow.

### CODE REVIEW

File: code/datums/gamemode/dynamic/dynamic.dm [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	for(var/mob/new_player/player in player_list)
+		if(player.ready && player.mind)
+			roundstart_pop_ready++
+			candidates.Add(player)
+	message_admins("Listing [roundstart_rules.len] round start rulesets, and [candidates.len] players ready.")
+	if (candidates.len <= 0)
+		message_admins("Not a single player readied-up. The round will begin without any roles assigned.")
+		return 1
+	if (roundstart_rules.len <= 0)
+		message_admins("There are no roundstart rules within the code, what the fuck? The round will begin without any roles assigned.")
```
Comment (MadmanMartian) [inline code/datums/gamemode/dynamic/dynamic.dm:50]: Keep it formal. No whadafuk in the user feedback

Comment (Exxion) [reply code/datums/gamemode/dynamic/dynamic.dm:50]: This is not a policy we have.

Comment (DeityLink) [reply code/datums/gamemode/dynamic/dynamic.dm:50]: It's an admin-only message, and it's not a professional game.

**Comment by DeityLink** [comment]
Autotraitor rulesets integrated. This should serve as a nice example of how they work for everyone.

File: code/datums/gamemode/dynamic/dynamic_rulesets_roundstart.dm [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		newTraitor.AssignToRole(M.mind,1)
+		newTraitor.OnPostSetup(FALSE)
+		newTraitor.Greet(GREET_ROUNDSTART)
+	return 1
+
+/datum/dynamic_ruleset/roundstart/traitor/update()
+	if (autotraitor_cooldown)
+		autotraitor_cooldown--
+	else
+		autotraitor_cooldown = 9000
```
Comment (MadmanMartian) [inline code/datums/gamemode/dynamic/dynamic_rulesets_roundstart.dm]: why not autotraitor_cooldown = initial(autotraitor_cooldown)?

File: code/datums/gamemode/dynamic/dynamic_rulesets_roundstart.dm [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+/datum/dynamic_ruleset/roundstart/traitor
+	name = "Syndicate Traitors"
+	persistent = 1
+	role_category = ROLE_TRAITOR
+	restricted_from_jobs = list("Cyborg","Mobile MMI","Security Officer", "Warden", "Detective", "Head of Security", "Captain")
+	required_candidates = 1
+	weight = 5
+	cost = 10
+	requirements = list(40,30,20,10,10,10,10,10,10,10)
+	var/autotraitor_cooldown = 9000
```
Comment (MadmanMartian) [inline code/datums/gamemode/dynamic/dynamic_rulesets_roundstart.dm]: Is this in ticks? If so, probably best for it to be 900 SECONDS for ease

**Comment by MadmanMartian** [comment]
Huh, turns out I was looking at an old commit. Thanks github

**Pull request merged**